src/librbd/crypto/luks/EncryptionFormat.h

```diff
@@ -18,29 +18,40 @@ template <typename ImageCtxT>
 class EncryptionFormat : public crypto::EncryptionFormat<ImageCtxT> {
 
 public:
+    EncryptionFormat(std::string&& passphrase);
     EncryptionFormat(encryption_algorithm_t alg, std::string&& passphrase);
     ~EncryptionFormat();
 
+    crypto::EncryptionFormat<ImageCtxT>* clone() const override {
+      std::string passphrase_copy = this->m_passphrase;
+      return new EncryptionFormat(std::move(passphrase_copy));
```

Comment:
It is indeed sneaky, with two purposes:
1. Preserve semantics for the current code in master.
2. Allow loading luks2 child with luks1 parent using qemu 6.1.0 (assuming same passphrase)

If we make it always discarded, then we don't preserve the semantics of the current master, which fails if e.g. user calls encryption_load("luks2",..) on a luks1 image.
If we make is always preserved, then the only way to load a luks2 child of a luks2 parent will be to use the more complex encryption_load2, which is not supported by qemu 6.1.0.

Current master actually complies (even if by error) with the suggested semantics here:
It only checks the child image, and blindly assumes the parent to be encrypted with the same key, with the same format (without even checking the parent header).
The suggested change only strengthens the check, is is verifies that the passphrase match for the parent.

In conclusion, if we insist on making it always preserve, we delay the use-case to when qemu will support encryption_load2.
Always discard will break current semantics, but will still allow the use-case via qemu 6.1.0.

Your call.

Reply:
> Current master actually complies (even if by error) with the suggested semantics here:
It only checks the child image, and blindly assumes the parent to be encrypted with the same key, with the same format (without even checking the parent header).

Well, current master gets a free pass because there is no real support for encrypted clones.

> Your call.

I would lean towards "always preserve".  I don't see the qemu use case you are referring to mattering at all -- who would encrypt one image in the chain with one format and another image in the chain with another format but using the same passphrase?  I can't think of any reason to do that, can you?

Reply:
> > Current master actually complies (even if by error) with the suggested semantics here:
> > It only checks the child image, and blindly assumes the parent to be encrypted with the same key, with the same format (without even checking the parent header).
> 
> Well, current master gets a free pass because there is no real support for encrypted clones.
> 
> > Your call.
> 
> I would lean towards "always preserve". I don't see the qemu use case you are referring to mattering at all -- who would encrypt one image in the chain with one format and another image in the chain with another format but using the same passphrase? I can't think of any reason to do that, can you?

Well we actually discussed this use-case on a call a couple of months ago.
It comes from spawning a child image of a LUKS1 exported image.

This use-case, along with the "auto-detect plaintext" use-case is the two sneaky features I'm trying to get so that they will work with current qemu, for client timeline reasons.

Maybe I'm shooting myself in the leg now, but I think that the 2 use-cases kind-of conform to "preserve only on child, discard on ancestors", as the plaintext parent use-case also does not preserve the rbd_ecnryption_format given by the user for that ancestor image being detected as plaintext.
Same here, we detect an ancestor as a different type of LUKS, and don't require it to be the same version.

Reply:
> Well we actually discussed this use-case on a call a couple of months ago. It comes from spawning a child image of a LUKS1 exported image.

Sorry, I don't recall that discussion -- maybe I'm confusing it with something else.  I understand that both cloning the format and assuming plaintext features are there for mostly for compatibility.  What I don't understand is why would someone want to change the format but keep the passphrase?

With current master, all clones have the same format and the same passphrase.  Going the "always preserve" route maintains that -- we either clone the specified format or fall back to plaintext -- the old `rbd_encryption_load` API still works.  But it sounds like you are pursuing an additional use case for `rbd_encryption_load` -- where someone, using the new Ceph bits, would create a clone with a different format but the same passphrase as the parent.  And this is where I'm getting stuck...  Isn't the point of this entire effort to support different passphrases?

> 
> This use-case, along with the "auto-detect plaintext" use-case is the two sneaky features I'm trying to get so that they will work with current qemu, for client timeline reasons.

For the purposes of this discussion, current QEMU means the old `rbd_encryption_load` API and nothing more, right?  Or am I just completely misunderstanding what "spawning a child image of a LUKS1 exported image" means?

Reply:
Right, current qemu = old rbd_encryption_load.

IMO plaintext is just a special case for a format.
So if you allow auto-detecting of it in rbd_encryption_load, why not allow other formats as well?

And to be more exact the use-case is migration from LUKS1 to LUKS2.
We actually want it to be transparent to the user, so of course passphrase stays the same.
You clone a LUKS1 image, format it to LUKS2 (with the same passphrase).
To fully transition, you can flatten the image, but it should still allow the user to use it while flattening (assuming flatten-copyups are guarded with CREATE flag for objects).

Reply:
> Right, current qemu = old rbd_encryption_load.
> 
> IMO plaintext is just a special case for a format. So if you allow auto-detecting of it in rbd_encryption_load, why not allow other formats as well?

I disagree.  First, we are not really auto-detecting it, we are falling back to it.  Second, what if we had a third, non-LUKS, format -- would you want to re-interpret a cloned LUKS1/LUKS2/LUKS-ANY spec as that format too?

> 
> And to be more exact the use-case is migration from LUKS1 to LUKS2. We actually want it to be transparent to the user, so of course passphrase stays the same.

That seems weird to me.  What does changing the format underneath the user (without them even knowing, I assume) buy?  Specifically, what does LUKS1 -> LUKS2 give you?

Is the user involved in picking the encryption format in the first place?

I'm not trying to nit pick here -- just really curious.

Reply:
You said it was my call so I expressed my preference for the "always preserve" option, but, if this use case is important to you and really can't wait/is blocked on QEMU integration, the "always discard" option is acceptable too.  It would change the current master semantics but it is better than inconsistent handling of the top-most image and the rest of the images in the chain.  Here is what I'm thinking would do there:

`rbd_encryption_format`
accept `LUKS1`, `LUKS2`, `SOME_FUTURE_FORMAT`
all are distinctive

`rbd_encryption_load` and `rbd_encryption_load2`
accept `LUKS1`, `LUKS2`, `LUKS`, `SOME_FUTURE_FORMAT`
`LUKS1`, `LUKS2`, `LUKS` are treated the same way
`SOME_FUTURE_FORMAT` is distinctive

Reply:
> That seems weird to me. What does changing the format underneath the user (without them even knowing, I assume) buy? Specifically, what does LUKS1 -> LUKS2 give you?

LUKS2 is faster and more CPU efficient than LUKS1 due to 4K sector size instead of 512 bytes.
For librbd end-to-end this gives 10-20% better performance.
Benchmarking just the encryption (with openssl speed) show even more of course.
> 
> Is the user involved in picking the encryption format in the first place?
> 
> I'm not trying to nit pick here -- just really curious.

Nope.

Reply:
Do we still need to add LUKS if it is (from rbd_encryption_load perspective) same as the already existing LUKS1 and LUKS2?

Reply:
If the user is not involved in picking the encryption format and you, knowing that LUKS2 is more efficient, default to LUKS2 for everyone, where would LUKS1-encrypted images even come from?

> Do we still need to add LUKS if it is (from rbd_encryption_load perspective) same as the already existing LUKS1 and LUKS2?

I think so, for the sake of clarity and correspondence to libcryptsetup's `CRYPT_LUKS` constant if nothing else.  Existing LUKS1/LUKS2 constants and "luks1"/"luks2" CLI option values can be marked deprecated for use on the loading side.

Reply:
> If the user is not involved in picking the encryption format and you, knowing that LUKS2 is more efficient, default to LUKS2 for everyone, where would LUKS1-encrypted images even come from?
> 

Mostly from use-cases which used QEMU LUKS1 encryption over RBD (at the time where LUKS2 was not available via librbd and QEMU). Now since LUKS2 is available we would like to migrate those "old" RBD images to the more efficient LUKS2.